pkg/controllers/provisioning/scheduling/instance_selection_test.go

```diff
@@ -0,0 +1,453 @@
+		ExpectInstancesWithLabel(cloudProv.CreateCalls[0].InstanceTypes, v1.LabelArchStable, "amd64")
+	})
+	It("should schedule on one of the cheapest instances (pod ct = spot/test-zone-2/amd64/linux)", func() {
+		pod := ExpectProvisioned(ctx, env.Client, selectionController, provisioners, provisioner, test.UnschedulablePod(
+			test.PodOptions{NodeRequirements: []v1.NodeSelectorRequirement{
+				{
+					Key:      v1.LabelArchStable,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{v1alpha5.ArchitectureAmd64},
+				},
+				{
+					Key:      v1.LabelOSStable,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{"linux"},
+				},
+				{
+					Key:      v1alpha5.LabelCapacityType,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{v1alpha1.CapacityTypeSpot},
+				},
+				{
+					Key:      v1.LabelTopologyZone,
+					Operator: v1.NodeSelectorOpIn,
+					Values:   []string{"test-zone-2"},
+				},
+			}}))
+		node := ExpectScheduled(ctx, env.Client, pod[0])
+		Expect(nodePrice(node)).To(Equal(minPrice))
+		ExpectInstancesWithOffering(cloudProv.CreateCalls[0].InstanceTypes, v1alpha1.CapacityTypeSpot, "test-zone-2")
+		ExpectInstancesWithLabel(cloudProv.CreateCalls[0].InstanceTypes, v1.LabelOSStable, "linux")
+		ExpectInstancesWithLabel(cloudProv.CreateCalls[0].InstanceTypes, v1.LabelArchStable, "amd64")
+	})
+	It("should schedule on an instance with enough resources", func() {
+		// these values are constructed so that three of these pods can always fit on at least one of our instance types
+		for _, cpu := range []float64{0.1, 1.0, 2, 2.5, 4, 8, 16} {
+			for _, mem := range []float64{0.1, 1.0, 2, 4, 8, 16, 32} {
+				cloudProv.CreateCalls = nil
+				opts := test.PodOptions{
+					ResourceRequirements: v1.ResourceRequirements{Requests: map[v1.ResourceName]resource.Quantity{
+						v1.ResourceCPU:    resource.MustParse(fmt.Sprintf("%0.1f", cpu)),
+						v1.ResourceMemory: resource.MustParse(fmt.Sprintf("%0.1fGi", mem)),
+					}}}
+				pods := ExpectProvisioned(ctx, env.Client, selectionController, provisioners, provisioner,
+					test.UnschedulablePod(opts), test.UnschedulablePod(opts), test.UnschedulablePod(opts))
+				nodeNames := sets.NewString()
+				for _, p := range pods {
+					node := ExpectScheduled(ctx, env.Client, p)
+					nodeNames.Insert(node.Name)
+				}
+				// should fit on one node
+				Expect(nodeNames).To(HaveLen(1))
+				totalPodResources := resources.RequestsForPods(pods...)
+				for _, it := range cloudProv.CreateCalls[0].InstanceTypes {
+					totalReserved := resources.Merge(totalPodResources, it.Overhead())
+					// the total pod resources in CPU and memory + instance overhead should always be less than the
+					// resources available on every viable instance has
+					Expect(totalReserved.Cpu().Cmp(it.Resources()[v1.ResourceCPU])).To(Equal(-1))
+					Expect(totalReserved.Memory().Cmp(it.Resources()[v1.ResourceMemory])).To(Equal(-1))
+				}
+			}
+		}
+	})
```

Comment:
If I understand correctly, the above cases can all find compatible instance type for the give pod. Can we add some cases with multiple pods that will create incompatibility and we should see them been put on separated nodes?

Reply:
This set of tests was geared at two recent issues that caused point releases where we were passing a set of instance types to the cloudprovider where some, but not all of the instance types matched the constraints. We have some individual tests for that behavior:
- "should launch pods with resources that aren't on any single instance type on different instances"
- "should schedule incompatible pods to the different node"